controllers/cloudstackfailuredomain_controller_test.go

```diff
@@ -49,14 +48,200 @@ var _ = Describe("CloudStackFailureDomainReconciler", func() {
 					arg1.(*infrav1.CloudStackZoneSpec).Network.Type = cloud.NetworkTypeShared
 				}).MinTimes(1)
 
-			tempfd := &infrav1.CloudStackFailureDomain{}
+		})
+
+		It("Should set failure domain Status.Ready to true.", func() {
+			assertFailureDomainCreated()
+		})
+		It("Should delete failure domain if no VM under this failure domain.", func() {
+			assertFailureDomainCreated()
+			Ω(k8sClient.Delete(ctx, dummies.CSFailureDomain1))
+
+			assertFailureDomainNotExisted()
+		})
+
+		It("Should return error if spec.replicas < 2.", func() {
+			assertFailureDomainCreated()
+			var specReplicas int32 = 1
+			var statusReplicas int32 = 1
+			var statusReadyReplicas int32 = 1
+			var statusReady = true
+			setCSMachineOwnerCRD(dummies.CSMachineOwner, &specReplicas, &statusReplicas, &statusReadyReplicas, &statusReady)
+			setCAPIMachineAndCSMachineCRDs(dummies.CSMachine1, dummies.CAPIMachine)
+			setMachineOwnerReference(dummies.CSMachine1, dummies.CSMachineOwnerReference)
+			labelMachineFailuredomain(dummies.CSMachine1, dummies.CSFailureDomain1)
+
+			Ω(k8sClient.Delete(ctx, dummies.CSFailureDomain1))
+
+			CAPIMachine := &clusterv1.Machine{}
+			Eventually(func() bool {
+				key := client.ObjectKey{Namespace: dummies.ClusterNameSpace, Name: dummies.CAPIMachine.Name}
+				if err := k8sClient.Get(ctx, key, CAPIMachine); err == nil {
+					return CAPIMachine.DeletionTimestamp.IsZero()
+				}
+				return false
+			}, timeout).WithPolling(pollInterval).Should(BeTrue())
+		})
+
+		It("Should return error if status.replicas < spec.replicas.", func() {
+			assertFailureDomainCreated()
+			var specReplicas int32 = 2
+			var statusReplicas int32 = 1
+			var statusReadyReplicas int32 = 1
+			var statusReady = true
+			setCSMachineOwnerCRD(dummies.CSMachineOwner, &specReplicas, &statusReplicas, &statusReadyReplicas, &statusReady)
+			setCAPIMachineAndCSMachineCRDs(dummies.CSMachine1, dummies.CAPIMachine)
+			setMachineOwnerReference(dummies.CSMachine1, dummies.CSMachineOwnerReference)
+			labelMachineFailuredomain(dummies.CSMachine1, dummies.CSFailureDomain1)
+
+			Ω(k8sClient.Delete(ctx, dummies.CSFailureDomain1))
+
+			CAPIMachine := &clusterv1.Machine{}
+			Eventually(func() bool {
+				key := client.ObjectKey{Namespace: dummies.ClusterNameSpace, Name: dummies.CAPIMachine.Name}
+				if err := k8sClient.Get(ctx, key, CAPIMachine); err == nil {
+					return CAPIMachine.DeletionTimestamp.IsZero()
+				}
+				return false
+			}, timeout).WithPolling(pollInterval).Should(BeTrue())
+		})
+
+		It("Should return error if status.ready is false.", func() {
```

Comment:
With the number of spec inputs here, I'm pretty sure these specs could be condensed into a table.

https://onsi.github.io/ginkgo/#table-specs

Would probably make the whole set of specs much clearer and greatly reduce code repetition.